engine/src/main/java/org/kigalisim/engine/SingleThreadEngine.java

```diff
@@ -456,39 +762,342 @@ private UnitConverter createUnitConverterWithTotal(String stream) {
+    BigDecimal kgForRecharge = rechargeVolume.getValue();
+    BigDecimal kgForNew = volumeForNew.getValue();
+
+    // Return to original initial charge
+    stateGetter.clearAmortizedUnitVolume();
+
+    // Return original
+    stateGetter.clearVolume();
+
+    // Determine how much to offset domestic and imports
+    EngineNumber manufactureRaw = getStream("manufacture", scopeEffective, null);
+    EngineNumber importRaw = getStream("import", scopeEffective, null);
+    EngineNumber priorRecycleRaw = getStream("recycle", scopeEffective, null);
+
+    EngineNumber manufactureSalesConverted = unitConverter.convert(manufactureRaw, "kg");
+    EngineNumber importSalesConverted = unitConverter.convert(importRaw, "kg");
+    EngineNumber priorRecycleSalesConverted = unitConverter.convert(priorRecycleRaw, "kg");
+
+    BigDecimal manufactureSalesKg = manufactureSalesConverted.getValue();
+    BigDecimal importSalesKg = importSalesConverted.getValue();
+    BigDecimal priorRecycleSalesKg = priorRecycleSalesConverted.getValue();
+    BigDecimal totalNonRecycleKg = manufactureSalesKg.add(importSalesKg);
+
+    // Get stream percentages for allocation
+    BigDecimal percentManufacture;
+    BigDecimal percentImport;
+    
+    if (totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0) {
+      EngineNumber manufactureInitialCharge = getInitialCharge("manufacture");
+      EngineNumber importInitialCharge = getInitialCharge("import");
+      BigDecimal manufactureInitialChargeVal = manufactureInitialCharge.getValue();
+      BigDecimal importInitialChargeVal = unitConverter
+          .convert(importInitialCharge, manufactureInitialCharge.getUnits()).getValue();
+      BigDecimal totalInitialChargeVal = manufactureInitialChargeVal.add(importInitialChargeVal);
+      
+      if (totalInitialChargeVal.compareTo(BigDecimal.ZERO) == 0) {
+        percentManufacture = BigDecimal.ONE;
+        percentImport = BigDecimal.ZERO;
+      } else {
+        percentManufacture = manufactureInitialChargeVal.divide(totalInitialChargeVal, 10, java.math.RoundingMode.HALF_UP);
+        percentImport = importInitialChargeVal.divide(totalInitialChargeVal, 10, java.math.RoundingMode.HALF_UP);
+      }
+    } else {
+      percentManufacture = totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0 
+          ? BigDecimal.ZERO : manufactureSalesKg.divide(totalNonRecycleKg, 10, java.math.RoundingMode.HALF_UP);
+      percentImport = totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0 
+          ? BigDecimal.ZERO : importSalesKg.divide(totalNonRecycleKg, 10, java.math.RoundingMode.HALF_UP);
+    }
+
+    // Recycle
+    EngineNumber newRecycleValue = new EngineNumber(recycledDisplacedKg, "kg");
+    this.streamKeeper.setStream(application, substance, "recycle", newRecycleValue);
+
+    // New values
+    BigDecimal requiredKgUnbound = kgForRecharge.add(kgForNew);
+    BigDecimal requiredKg = requiredKgUnbound.compareTo(BigDecimal.ZERO) < 0 ? BigDecimal.ZERO : requiredKgUnbound;
```

Comment:
Please put `requiredKgUnbound.compareTo(BigDecimal.ZERO) < 0` into `requiredKgNegative` to simplify.